### PR TITLE
Performance Improvements When Parametrizing Systems

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description
This PR aims to fix slow performance when using the toolkit to parametrize systems, caused by many calls to the `TopologyMolecule.xxx_start_topology_index` and the `TopologyParticle.topology_xxx_index` methods.

As an example of the speed up of this PR, 

- the **unoptimized** times for parametrizing 1 and 1000 triethanolamine molecules are ~1470ms and ~521360ms respectively and for parametrizing 1 and 1000 benzene molecules are ~300ms and ~265700ms using the 0.3.0 version of the toolkit from conda.
- the **optimized** times for parametrizing 1 and 1000 triethanolamine molecules are ~1060ms and ~16930ms respectively and for parametrizing 1 and 1000 benzene molecules are ~320ms and ~8720ms using this PR branch.

so roughly a 30x speed up for the 1000 molecule systems (the performance gain is expected to be even larger for bigger systems).

While this is definitely an improvement, I would recommend perhaps having a rethink of the `Topology` classes generator design pattern - while it should be memory efficient, in the cases I've looked at here there can be a significant compute overhead associated with it. I think a compromise between a fully computed `Topology` object, and a when needed generator design is definitely possible, but will likely require a rewrite of the `Topology` classes.

## TODOS
- [x] Tag issue being addressed
- [x] Update [tests](https://github.com/openforcefield/openforcefield/tree/master/openforcefield/tests)
- [x] Update docstrings/[documentation](https://github.com/openforcefield/openforcefield/tree/master/docs), if applicable
- [ ] Update [changelog](https://github.com/openforcefield/openforcefield/blob/master/docs/releasehistory.rst)

## Status
- [x] Ready to review
- [ ] Ready to go

### CODE REVIEW

**Comment by jchodera** [comment]
> While this is definitely an improvement, I would recommend perhaps having a rethink of the Topology classes generator design pattern - while it should be memory efficient, in the cases I've looked at here there can be a significant compute overhead associated with it. I think a compromise between a fully computed Topology object, and a when needed generator design is definitely possible, but will likely require a rewrite of the Topology classes.

Would be great if you could capture this insight/suggestion in a separate issue!

**Comment by j-wags** [comment]
Excellent -- I'm deep in some 0.3 spec stuff, but will review this later today!

**Comment by SimonBoothroyd** [comment]
> Would be great if you could capture this insight/suggestion in a separate issue!

Sure! I added this onto the discussion in issue #253

File: openforcefield/topology/molecule.py [hunk 1]
```diff
@@ -3317,6 +3317,44 @@ def _is_bonded(self, atom_index_1, atom_index_2):
...
+
+        """
+        if isinstance(i, int) and isinstance(j, int):
+            atom_i = self._atoms[i]
+            atom_j = self._atoms[j]
+        elif isinstance(i, Atom) and isinstance(j, Atom):
+            atom_i = i
+            atom_j = j
+        else:
+            raise Exception(
```
[the remark this answers is not in the document]
Comment (j-wags) [reply openforcefield/topology/molecule.py]: Updated in my commit.

[the remark this answers is not in the document]
Comment (j-wags) [reply openforcefield/topology/molecule.py]: Ugh. I used BaseExceptions all over the place when I wrote this. *Sigh*

**Pull request merged**